Rewrite the non-Azure messaging integration articles (#1759)
dotnet/docs-aspire: docs/messaging/nats-integration.md

title: .NET Aspire NATS integration
description: Learn how to use the .NET Aspire NATS integration to send logs and traces to a NATS Server.

> ms.topic: how-to

Reviewer: All of these integration articles read like conceptual or overview mixes. Maybe you should change the topic type to conceptual?

Author: Yes, will do. I like that idea.

ms.date: 10/09/2024
---

# .NET Aspire NATS integration

[!INCLUDE [includes-hosting-and-client](../includes/includes-hosting-and-client.md)]

[NATS](https://nats.io) is a high-performance, secure, distributed messaging system. The .NET Aspire NATS integration enables you to connect to existing NATS instances, or create new instances from .NET with the [`docker.io/library/nats` container image](https://hub.docker.com/_/nats).

## Hosting integration

> The NATS hosting integration models a NATS server as the <xref:Aspire.Hosting.ApplicationModel.NatsServerResource> type. To access this type and APIs that allow you to add it to your app host, install the [📦 Aspire.Hosting.Nats](https://www.nuget.org/packages/Aspire.Hosting.Nats) NuGet package in the [app host](xref:aspire/app-host) project.

Reviewer: You say app host twice in one sentence, and I don't feel like it flows well. This format is repeated for all of the integrations. I also don't understand the terminology super well, so this is just a suggestion and feel free to ignore it.

```suggestion
NATS hosting integration for .NET Aspire models a NATS server as the <xref:Aspire.Hosting.ApplicationModel.NatsServerResource> type. To access this type, install the [📦 Aspire.Hosting.Nats](https://www.nuget.org/packages/Aspire.Hosting.Nats) NuGet package in the [app host](xref:aspire/app-host) project, then add it with the builder.
```

### [.NET CLI](#tab/dotnet-cli)

```dotnetcli
dotnet add package Aspire.Hosting.Nats
```

### [PackageReference](#tab/package-reference)

```xml
<PackageReference Include="Aspire.Hosting.Nats"

>                   Version="[SelectVersion]" />

Reviewer: Is there a specific reason you have `[SelectVersion]`? If you're just saying "the latest," use `*` instead.

```

---

For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-package) or [Manage package dependencies in .NET applications](/dotnet/core/tools/dependencies).

### Add NATS server resource

In your app host project, call <xref:Aspire.Hosting.NatsBuilderExtensions.AddNats*> on the `builder` instance to add a NATS server resource:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var nats = builder.AddNats("nats");

builder.AddProject<Projects.ExampleProject>()
       .WithReference(nats);

// After adding all resources, run the app...
```

> When .NET Aspire adds a container image to the app host, as shown in the preceding example with the `docker.io/library/nats` image, it creates a new NATS server instance on your local machine. A reference to your NATS server (the `nats` variable) is added to the `ExampleProject`.

Reviewer: I'm not sure what preceding example you're talking about that references a docker image? Is there a default image that's being applied or something?

Author: The code right before this text. The call to `AddNats` creates a NATS server container instance on your local machine.

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var nats = builder.AddNats("nats");
                  .WithJetStream();

builder.AddProject<Projects.ExampleProject>()
       .WithReference(nats);

// After adding all resources, run the app...
```

The NATS JetStream functionality provides a built-in persistence engine called JetStream which enables messages to be stored and replayed at a later time. You can optionally provide a `srcMountPath` parameter to specify the path to the JetStream data directory on the host machine (the provided mount path maps to the container's `-sd` argument).

### Add NATS server resource with data volume

To add a data volume to the NATS server resource, call the <xref:Aspire.Hosting.NatsBuilderExtensions.WithDataVolume*> method on the NATS server resource:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var nats = builder.AddNats("nats");
                  .WithDataVolume(isReadOnly: false);

builder.AddProject<Projects.ExampleProject>()
       .WithReference(nats);

// After adding all resources, run the app...
```

> The data volume is used to persist the NATS server data outside the lifecycle of its container. The data volume is mounted at the `/var/lib/nats` path in the NATS server container and when a `name` parameter isn't provided, the name is generated at random. For more information on data volumes and details on why they're preferred over [bind mounts](#add-nats-server-resource-with-data-bind-mount), see [Docker docs: Volumes](https://docs.docker.com/engine/storage/volumes).

Reviewer: Nit: feels like a run on sentence. I swapped "name at random" with "provide the parameter" assuming that the more common scenario is a randomly generated name, considering your example didn't provide a name. Correct?

```suggestion
The data volume is used to persist the NATS server data outside the lifecycle of its container. The data volume is mounted at the `/var/lib/nats` path in the NATS server container. A name is generated at random unless you provide a set the `name` parameter. For more information on data volumes and details on why they're preferred over [bind mounts](#add-nats-server-resource-with-data-bind-mount), see [Docker docs: Volumes](https://docs.docker.com/engine/storage/volumes).
```

### Add NATS server resource with data bind mount

To add a data bind mount to the NATS server resource, call the <xref:Aspire.Hosting.NatsBuilderExtensions.WithDataBindMount*> method:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var nats = builder.AddNats("nats");
                  .WithDataBindMount(
                      source: "/NATS/Data",
                      isReadOnly: false);

builder.AddProject<Projects.ExampleProject>()
       .WithReference(nats);

// After adding all resources, run the app...
```

> 

Reviewer: Is this a good place to insert a TIP or WARNING about why they should use a data volume instead?

Author: Done

```xml
<PackageReference Include="Aspire.NATS.Net"
                  Version="[SelectVersion]" />
```

---

### Add NATS client

In the _:::no-loc text="Program.cs":::_ file of your client-consuming project, call the <xref:Microsoft.Extensions.Hosting.AspireNatsClientExtensions.AddNatsClient*> extension method on any <xref:Microsoft.Extensions.Hosting.IHostApplicationBuilder> to register an `INatsConnection` for use via the dependency injection container. The method takes a connection name parameter.

```csharp
builder.AddNatsClient(connectionName: "nats");
```

> [!TIP]
> The `connectionName` parameter must match the name used when adding the NATS server resource in the app host project. For more information, see [Add NATS server resource](#add-nats-server-resource).

You can then retrieve the `INatsConnection` instance using dependency injection. For example, to retrieve the client from a service:

```csharp
public class ExampleService(INatsConnection connection)
{
    // Use connection...
}
```

For more information on dependency injection, see [.NET dependency injection](/dotnet/core/extensions/dependency-injection).

### Add keyed NATS client

> There might be situations where you want to register multiple `INatsConnection` instances with different connection names. To register keyed NATS clients, call the <xref:Microsoft.Extensions.Hosting.AspireNatsClientExtensions.AddKeyedNatsClient*>:

Reviewer: ```suggestion
There might be situations where you want to register multiple `INatsConnection` instances with different connection names. To register keyed NATS clients, call the <xref:Microsoft.Extensions.Hosting.AspireNatsClientExtensions.AddKeyedNatsClient*> method:
```
